src/lsm/table_memory.zig

```diff
@@ -136,23 +148,102 @@ pub fn TableMemoryType(comptime Table: type) type {
-                .mutability = .mutable,
+                .mutability = .{ .mutable = .{} },
                 .name = table.name,
             };
         }
 
         pub fn sort(table: *TableMemory) void {
+            assert(table.mutability == .mutable);
+
             if (!table.value_context.sorted) {
-                std.mem.sort(
-                    Value,
-                    table.values_used(),
-                    {},
-                    sort_values_by_key_in_ascending_order,
-                );
+                table.sort_suffix_from_offset(0);
                 table.value_context.sorted = true;
             }
         }
 
+        pub fn sort_suffix(table: *TableMemory) void {
+            assert(table.mutability == .mutable);
+            assert(table.mutability.mutable.suffix_offset <= table.count());
+
+            table.sort_suffix_from_offset(table.mutability.mutable.suffix_offset);
+
+            assert(table.mutability.mutable.suffix_offset == table.count());
+        }
+
+        fn sort_suffix_from_offset(table: *TableMemory, offset: u32) void {
+            assert(table.mutability == .mutable);
+            assert(offset == table.mutability.mutable.suffix_offset or offset == 0);
+            assert(offset <= table.count());
+
+            std.mem.sort(
+                Value,
+                table.values_used()[offset..],
+                {},
+                sort_values_by_key_in_ascending_order,
+            );
+
+            // Merge values with identical keys (last one wins) and collapse tombstones for
+            // secondary indexes.
+            const source_count: u32 = table.count();
+            var source_index: u32 = offset;
+            var target_index: u32 = offset;
+            while (source_index < source_count) {
+                table.values[target_index] = table.values[source_index];
+
+                // If we're at the end of the source, there is no next value, so the next value
+                // can't be equal.
+                const value_next_equal = source_index + 1 < source_count and
+                    key_from_value(&table.values[source_index]) ==
+                    key_from_value(&table.values[source_index + 1]);
+
+                if (value_next_equal) {
+                    if (Table.usage == .secondary_index) {
+                        // Secondary index optimization --- cancel out put and remove.
+                        // NB: while this prevents redundant tombstones from getting to disk, we
+                        // still spend some extra CPU work to sort the entries in memory. Ideally,
+                        // we annihilate tombstones immediately, before sorting, but that's tricky
+                        // to do with scopes.
+                        assert(Table.tombstone(&table.values[source_index]) !=
+                            Table.tombstone(&table.values[source_index + 1]));
+                        source_index += 2;
+                        target_index += 0;
```

Comment:
We now run this logic only for chunks. That is, if there's an in-del pair that strides a chunk boundary, we won't collapse it. I _think_ it is correct, but could you add an assert that the value at `offset-1` is not collapsible

Reply:
> We now run this logic only for chunks.

It runs for chunks initially, but then `make_immutable()` calls `.sort()` which in turn calls `sort_suffix_from_offset(0);`.